server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -298,35 +299,95 @@ public void parse(ParseContext context) throws IOException {
      * @param format an optional format string used when formatting values, for example a date format.
      * @return a list a standardized field values.
      */
-    public List<?> lookupValues(SourceLookup lookup, @Nullable String format) {
-        Object sourceValue = lookup.extractValue(name(), nullValue());
-        if (sourceValue == null) {
-            return List.of();
-        }
+    public abstract ValueFetcher valueFetcher(SearchLookup lookup, @Nullable String format);
```

Comment:
I decided to go "leaf by leaf" here because that is the kind of iterations that you'd need to read doc values runtime fields.  Another option would be to make fetch take both the `LeafReaderContext` and the `docId` and have the doc values and runtime fields based implementations build their own leaf readers. That feels worse to me but it is certainly an option.